Schema.org offer markup changes.

leewillis77 opened the pull request:
This PR is a first attempt to address the schema.org markup issues raised in #11281. The PR contains the following changes:
- Schema.org markup only output for simple products, variable products, and external products
- No markup output for grouped products
- Markup attached to the woocommerce_*_add_to_cart hooks, no longer part of price template
- New template files added under single-product/schema-offers/
- Multiple offers output for variable products, one per available variation

There are probably still some issues to be worked through before this finally goes in, my thoughts so far are:
- Probably needs to go into a major release due to new template files?
- Can we output sensible schema markup for grouped products?
- What (if anything) do we need to do for Composite products and any other similar add-ons - do they currently generate schema markup?

So - PR for discussion at this stage

leewillis77 committed:
Schema.org offer markup changes.

This contains the following changes:
- Schema.org markup only output for simple products, variable products, and external products
- No markup output for grouped products
- Markup attached to the woocommerce_*_add_to_cart hooks, no longer part of price template
- New template files added under single-product/schema-offers/
- Multiple offers output for variable products, one per available variation

leewillis77 commented:
Output from Google's structure data testing tool below:

**Simple product:**
<img width="704" alt="simple" src="https://cloud.githubusercontent.com/assets/1097338/16764305/3687f60a-4825-11e6-9161-5ccecabdeea4.png">

**External product:**
<img width="700" alt="external" src="https://cloud.githubusercontent.com/assets/1097338/16764314/42602cc2-4825-11e6-9281-ded047a84341.png">

**Variable product:**
<img width="684" alt="variable" src="https://cloud.githubusercontent.com/assets/1097338/16764320/48136530-4825-11e6-9c61-7f9343f18c0f.png">

opportus commented:
Does a dynamically generated JSON wouldn't be a much more flexible and modern way to implement shema.org now in WooCommerce ?
See what I'm talking about in this issue [#399](https://github.com/woothemes/storefront/issues/399) on Storefront repo, followed by this [commit](https://github.com/woothemes/storefront/commit/f05eea71176479a344ddf025b8e245ca383325cd).

franticpsyx commented:
>Can we output sensible schema markup for grouped products?
>What (if anything) do we need to do for Composite products and any other similar add-ons - do they currently generate schema markup?

How about https://schema.org/AggregateOffer ?

This bit would cause a big hit for large numbers of variations, when variations data is actually fetched via AJAX on the front-end:

https://github.com/woothemes/woocommerce/pull/11411/files#diff-65760bf45f136dba5d546afa2c8b7ee9R1039

Might make sense for variable products with few variations, though.

For big variable + other complex product types, it would be nice to use something like `AggregateOffer`, to provide min/max price data only, possibly along with a "base" SKU.

With Grouped/Bundle/Composite products it is impossible to come up with a list of all possible quantity+attribute/product combinations, at least not without a big hit on the server.

leewillis77 commented:
Hi @opportus; 

WooCommerce already implements schema via the Microdata approach rather than JSON-LD. This PR isn't aimed at changing that throughout, just addressing a specific issue (#11281) where the schema markup isn't complete / correct. If you want to suggest moving everything over to JSON-LD, then it should probably go as a new issue & PR - although Microdata is still valid, and supported by Google at the moment, so not sure what the benefit of changing would be.

leewillis77 commented:
Hi @franticpsyx 

Thanks for the comments - I've replied to your specific points below:

> How about https://schema.org/AggregateOffer ?
I considered it, but I don't think it'd give enough information to allow Google to verify individual prices.

> This bit would cause a big hit for large numbers of variations, when variations data is actually fetched via AJAX on the front-end:
> https://github.com/woothemes/woocommerce/pull/11411/files#diff-65760bf45f136dba5d546afa2c8b7ee9R1039
> 

Variation data isn't always fetched via AJAX though - right? Perhaps I should be checking woocommerce_ajax_variation_threshold and use that to determine whether to use aggregateOffer, similar to https://github.com/woothemes/woocommerce/blob/master/includes/wc-template-functions.php#L974 ?

> With Grouped/Bundle/Composite products it is impossible to come up with a list of all possible quantity+attribute/product combinations, at least not without a big hit on the server.

Understood - that's why I left it out at this stage - just wanted to be clear that it wasn't included.

opportus commented:
Hi @leewillis77 , I know I'm a bit out of topic here, but as you said yourself the actual shema.org markup is still valid and I don't know if opening an issue for that would be more appropriated since this is not, so I profit of this occasion here to speak about it, forgive me :p

For single product, one of the benefits of using JSON-LD over markup, would be that manipulating/modifying the default structured data would become much more easy and clean for other plugins or users. Just with the help of a single filter hook avoiding heavy changes in functions/templates...

See this [example](https://github.com/woothemes/storefront/commit/f05eea71176479a344ddf025b8e245ca383325cd#diff-76610de69674bf1428269d5a6993f8b8R359).

franticpsyx commented:
> Variation data isn't always fetched via AJAX though - right? Perhaps I should be checking woocommerce_ajax_variation_threshold and use that to determine whether to use aggregateOffer

+1

leewillis77 committed:
Support aggregateOffer for products with large number of variations.

Look at woocommerce_ajax_variation_threshold to decide if we should include
schema data for individual variations or not. Allow that choice to be
overridden by a new filter 'woocommerce_show_variation_schema_data' so that
people can continue to have variation data loaded via AJAX, but still have
detailed schema information output if they want.

Adds new template for aggregate offer output, plus associated hooks.

Also, rename variables in schema-offers/variation.php template for clarity.

leewillis77 commented:
@franticpsyx I updated the PR to support aggregateOffer if the number of variations is over the AJAX threshold. I can't validate the markup as Google's structured data testing tool is throwing a fit today and refusing to validate anything, the code is pretty simple so should be fine:

```html
<div itemprop="offers" itemscope itemtype="http://schema.org/AggregateOffer">
	<meta itemprop="priceCurrency" content="GBP" />
	<meta itemprop="lowPrice" content="15.00" />
	<meta itemprop="highPrice" content="17.00" />
	<meta itemprop="offerCount" content="4" />
</div>
```

franticpsyx commented:
@opportus 
>For single product, the main benefits of using JSON-LD over markup, would be that manipulating/modifying the default structured data would become much more easy and clean for other plugins or users. Just with the help of a single filter hook avoiding heavy changes in functions/templates... Also, the shema.org stuff would be centralized in a single function instead of being scattered all over and would get independant from templates.

I think this is some interesting feedback - however, this PR does decouple the structured data from the templates a bit, so it makes things easier already. I haven't worked with structured data or SEO a lot, so hopefully someone else can chime in and provide some extra insight on the pros/cons of each approach.

@leewillis77 
>I updated the PR to support aggregateOffer if the number of variations is over the AJAX threshold.

Looks good. I'm not quite sure what the `offerCount` field should ideally represent, though. So for instance, would a grouped product include as many offers as the different products it contains? Would the same be true for Composite products (for instance, X components by Y products = X*Y offers)? Or would the `offerCount` field contain all the different valid SKU combinations (disregarding quantities, which would actually change the "aggregate SKU).

leewillis77 commented:
@opportus

> For single product, the main benefits of using JSON-LD over markup, would be that manipulating/modifying the default structured data would become much more easy and clean for other plugins or users.

Agreed. I'd suggest opening a fresh issue for the switch to JSON-LD.

@franticpsyx 

> I'm not quite sure what the offerCount field should ideally represent, though.

It's the number of variations. 

> So for instance, would a grouped product include as many offers as the different products it contains? Would the same be true for Composite products (for instance, X components by Y products = X*Y offers)? Or would the offerCount field contain all the different valid SKU combinations (disregarding quantities, which would actually change the "aggregate SKU").

Grouped / composite products don't output any schema - they're excluded as per my original comment.

franticpsyx commented:
>Grouped / composite products don't output any schema - they're excluded as per my original comment.

Sure, but since you brought up this point I thought we might be able to come up with a meaningful solution for those types as well:

>What (if anything) do we need to do for Composite products and any other similar add-ons - do they currently generate schema markup?

opportus commented:
Hi @leewillis77 

Since you both found interesting the idea of switching to JSON-LD, I've done the PR #11455 for it.
Notice BTW how your changes (which I +1) seem to be much easier to implement on top of mines...

mikejolley commented:
See #11455

galapogos01 commented:
If there is going to be a patch release for 2.6, I hope this gets included!

claudiosanches commented:
@galapogos01 there is no more 2.6.x releases, need to update to 3.0 now.
